convert2rhel/subscription.py

```diff
@@ -120,33 +136,94 @@ def register_system():
+        loggerinst.info("Unregistering the system to clear the server's state for our registration.")
+        try:
+            unregister_system()
+        except UnregisterError as e:
+            loggerinst.warning(str(e))
+
+        # Hack: currently, on RHEL7, subscription-manager unregister is
+        # reporting that the system is not registered but then calling the
+        # subscription-manager Registration API reports that the system is
+        # already registered. We suspect that the test host is being used for
+        # a previous test.  In that test, the system is registered, then
+        # rollback occurs which unregisters the server in some places but it
+        # remains registered in other data.  Then registration fails.
+        #
+        # For the short term, we are going to stop the rhsm service to work
+        # around this issue.  It should restart on its own when needed:
+        loggerinst.info(
+            "Stopping the RHSM service so that registration does not think that the host is already registered."
+        )
+        try:
+            _stop_rhsm()
+        except StopRhsmError as e:
+            # The system really might not be registered yet and also not running rhsm
+            # so ignore the error and try to register the system.
+            loggerinst.info(str(e))
+
+        # Register the system
         registration_cmd = RegistrationCommand.from_tool_opts(tool_opts)
-        output, ret_code = registration_cmd()
-        if ret_code == 0:
-            # Handling a signal interrupt that was previously handled by
-            # subscription-manager.
-            if "user interrupted process" in output.lower():
-                raise KeyboardInterrupt
-            return
 
-        loggerinst.info("System registration failed with return code = %s" % str(ret_code))
-        sleep(REGISTRATION_ATTEMPT_DELAYS[attempt])
-        attempt += 1
-    loggerinst.critical("Unable to register the system through subscription-manager.")
+        try:
+            registration_cmd()
+            loggerinst.info("System registration succeeded.")
+        except KeyboardInterrupt:
+            # When the user hits Control-C to exit, we shouldn't retry
+            raise
+        except Exception as e:
+            loggerinst.info("System registration failed with error: %s" % str(e))
+            sleep(REGISTRATION_ATTEMPT_DELAYS[attempt])
+            attempt += 1
+            continue
+
+        break
+
+    else:  # While-else
+        # We made the maximum number of subscription-manager retries and still failed
+        loggerinst.critical("Unable to register the system through subscription-manager.")
+
+    return None
+
+
+def _stop_rhsm():
+    """Stop the rhsm service."""
+    cmd = ["/bin/systemctl", "stop", "rhsm"]
+    if system_info.version.major <= 6:
+        # RHEL6 used upstart.  After that, we use systemd
+        cmd = ["service", "rhsm", "stop"]
```

Comment:
```
[08/17/2022 20:01:59] DEBUG - Calling command 'service rhsm stop'
Stopping RHSM failed with code: 1; output: rhsm: unrecognized service
```

I looked at the services (`service --status-all`) and there's just `rhsmcertd` which I'm not sure makes sense stopping (service description: `Enable periodic update of entitlement certificates.`).
```
rpc.svcgssd is stopped
rpc.mountd is stopped
nfsd is stopped
rpc.statd (pid  1162) is running...
master (pid  1326) is running...
rdisc is stopped
restorecond is stopped
rhsmcertd is stopped
rpcbind (pid  1140) is running...
rpc.gssd is stopped
rpc.idmapd is stopped
rpc.svcgssd is stopped
rsyslogd (pid  1111) is running...
sandbox is stopped
saslauthd is stopped
openssh-daemon (pid  19422) is running...
```

Reply:
Hmmm.... Okay, so I just ran our dbus test script on CentOS6 and then looked at the process list and found this:
> 3485 ?        Sl     0:00 python /usr/libexec/rhsm-service

I think that rhsm-service is started on-demand when something uses the DBus API.  On systemd hosts (CentOS7 and up), this is handled by systemd which waits for a socket to be opened and then uses that to launch rhsm-service so that it can handle the request.  On centos6, the subscription-manager code must do this itself and they haven't made a SysVinit script to start the service there.  If we need to stop the service, I think we'll need to find the rhsm-service process and send it a `kill -TERM`.

I haven't been able to provoke the system is already registered error with the dbus reg.py script so I'm not certain whether RHEL6 has this problem or not.

Here's what I propose... I will take out the code path that operates on CentOS-6.  If you want to look into whether it's needed on CentOS 6 (or if we want to wait to see if a user runs into a problem) we can implement sending a signal to rhsm-service then.  Does that sound good?

Reply:
I was also thinking that on RHEL6-like systems we may not need to stop the service because it works differently there. So leaving the service-stopping out for OS 6 sounds good to me. Thanks for looking into it.